go/store/val/tuple_builder.go

```diff
@@ -80,12 +90,68 @@ func (tb *TupleBuilder) Build(pool pool.BuffPool) (tup Tuple) {
 			panic("cannot write NULL to non-NULL field")
 		}
 	}
-	return tb.BuildPermissive(pool)
+	return tb.BuildPermissive(pool, tb.vs)
 }
 
 // BuildPermissive materializes a Tuple from the fields
 // written to the TupleBuilder without validating nullability.
-func (tb *TupleBuilder) BuildPermissive(pool pool.BuffPool) (tup Tuple) {
+func (tb *TupleBuilder) BuildPermissive(pool pool.BuffPool, vs ValueStore) (tup Tuple) {
+	// DO NOT SUBMIT: add context parameter
+	ctx := context.Background()
+	// Values may get passed into the tuple builder in either in-band or out-of-band form.
+	// In the best case, we don't need to convert either of them. But we need to track the sizes.
+
+	// We need to determine which TOAST fields get written out, and pass the addresses in instead. This is the only
+	// place where we break encapsulation.
+	// Track the total packed and unpacked size?
+	totalSize := tb.inlineSize
+	if totalSize > tb.tupleLengthTarget {
+		// start allocating to toast fields until we get below the size.
+		// do this in a way that doesn't require the full field.
+		for i, descType := range tb.Desc.Types {
+			if IsToastEncoding(descType.Enc) {
+				toastValue := ToastValue(tb.fields[i])
+				outlineSize := toastValue.outlineSize()
+				inlineSize := toastValue.inlineSize()
+
+				if outlineSize < inlineSize {
+					// outline this field.
+					if !toastValue.IsOutlined() {
+						// This isn't correct. Can we avoid the extra copy cleanly? If not, don't prematurely optimize.
+						outline, err := toastValue.convertToOutline(ctx, tb.vs, nil)
+						if err != nil {
+							// TODO: return error
+							return nil
+						}
+						tb.PutRaw(i, outline)
+					}
+
+					totalSize += outlineSize - inlineSize
+				}
+			}
+
+			if totalSize <= int64(MaxTupleDataSize) {
+				// We have enough space, mark all the remaining columns as inline
+				for j, descType := range tb.Desc.Types[i+1:] {
+					if IsToastEncoding(descType.Enc) {
+						toastValue := ToastValue(tb.fields[j+i+1])
+						if !toastValue.isInlined() {
+							inline, err := toastValue.convertToInline(ctx, tb.vs, nil)
+							if err != nil {
+								return nil
+							}
+							tb.PutRaw(j+i+1, inline)
+						}
+					}
+				}
+				break
+			}
+		}
+
+	}
+	if totalSize > int64(MaxTupleDataSize) {
+		// Outlining failed to reduce the tuple size enough. Is this possible?
```

Comment:
It shouldn't be, but there should be checks that prevent it. We check schemas on creation to make sure that they cannot result in an oversize tuple, and that should be the case with these new encodings as well. Make sure to have tests.